### PR TITLE
impl(generator): refactor generated rest stub to use EndpointOption

### LINKED ISSUE #10214: Leverage Options to override rest transport endpoint
Current generated RestStubs hard code the endpoint from the proto file annotation. Modify the emitted code to check the passed Options for a possible override for the default endpoint.

### PR DESCRIPTION
**Pull request by scotthart**
fixes #10214

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10273)

### CODE REVIEW

File: generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc [hunk 1]
```diff
@@ -30,7 +31,7 @@ GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
 
 DefaultGoldenKitchenSinkRestStub::DefaultGoldenKitchenSinkRestStub(Options options)
     : rest_client_(rest_internal::MakePooledRestClient(
-          "https://goldenkitchensink.googleapis.com", options)),
+      options.get<EndpointOption>(), options)),
```
Comment (coryan) [inline generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc:34]: I think the default value for generated libraries is something like `compute.googleapis.com`, or `bigquery.googleapis.com`.  Do I get that right?
If so, do you need to add the `http://` prefix?

File: generator/internal/stub_rest_generator.cc [hunk 2]
```diff
@@ -152,11 +153,10 @@ Status StubRestGenerator::GenerateCc() {
   auto result = CcOpenNamespaces(NamespaceType::kInternal);
   if (!result.ok()) return result;
 
-  // TODO(#10214): Emit code to check options for a potential endpoint override.
   CcPrint(R"""(
 Default$stub_rest_class_name$::Default$stub_rest_class_name$(Options options)
     : rest_client_(rest_internal::MakePooledRestClient(
-          "https://$service_endpoint$", options)),
+      options.get<EndpointOption>(), options)),
```
Comment (devbww) [inline generator/internal/stub_rest_generator.cc:159]: Restore indentation?

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10273)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @coryan)

___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc` line 34 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10273#-NHBVvvPK2SIxDt1r57Q-r1-34:-NHBXb874x0tVTrVc9Fy:b-2m280b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/e8e79b5f64419360b9592be7eb6748a29be08f28/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc#L34)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

I think the default value for generated libraries is something like `compute.googleapis.com`, or `bigquery.googleapis.com`.  Do I get that right?
If so, do you need to add the `http://` prefix?
</blockquote></details>

You're correct about the default endpoint option value for generated libraries. I'm looking at whether or not I need to sanitize the option value (i.e. add the https:// if it's missing). And I'd prefer to do that in the RestClient rather than here. These types of runtime issues will be verified in the eventual integration tests running against some server.

File: generator/internal/stub_rest_generator.cc [hunk 3]
```diff
@@ -152,12 +153,11 @@ Status StubRestGenerator::GenerateCc() {
...
   if (!result.ok()) return result;
 
-  // TODO(#10214): Emit code to check options for a potential endpoint override.
   CcPrint(R"""(
 Default$stub_rest_class_name$::Default$stub_rest_class_name$(Options options)
     : rest_client_(rest_internal::MakePooledRestClient(
-          "https://$service_endpoint$", options)),
-      options_(std::move(options)) {}
+          options.get<EndpointOption>(), options)),
+          options_(std::move(options)) {}
```
Comment (devbww) [inline generator/internal/stub_rest_generator.cc:160]: It looks like you restored the one line, but then added indentation to the next.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10273)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @coryan and @devbww)

___
*[`generator/internal/stub_rest_generator.cc` line 159 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10273#-NHBVvvPK2SIxDt1r57S-r1-159:-NHBbUh97FflWuAN1aW4:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/e8e79b5f64419360b9592be7eb6748a29be08f28/generator/internal/stub_rest_generator.cc#L159)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Restore indentation?
</blockquote></details>

Done.
___
*[`generator/internal/stub_rest_generator.cc` line 160 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10273#-NHBVvvPK2SIxDt1r57S-r2-160:-NHBbUzi5h1s05AiLDZf:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9173c254cd288ebe03764178f53f01840f23cca3/generator/internal/stub_rest_generator.cc#L160)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

It looks like you restored the one line, but then added indentation to the next.
</blockquote></details>

Done.

**Review by coryan — APPROVED** [review]
> Reviewed 1 of 1 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10273)* status: all files reviewed, 2 unresolved discussions (waiting on @devbww)

___
*[`generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc` line 34 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/10273#-NHBVvvPK2SIxDt1r57Q-r1-34:-NHUaUSKCEN7PBVBW1Ez:b-svh06d) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/e8e79b5f64419360b9592be7eb6748a29be08f28/generator/integration_tests/golden/internal/golden_kitchen_sink_rest_stub.cc#L34)):*
<details><summary><i>Previously, scotthart (Scott Hart) wrote…</i></summary><blockquote>

You're correct about the default endpoint option value for generated libraries. I'm looking at whether or not I need to sanitize the option value (i.e. add the https:// if it's missing). And I'd prefer to do that in the RestClient rather than here. These types of runtime issues will be verified in the eventual integration tests running against some server.
</blockquote></details>

Hmm... Okay, maybe it will work.

**Pull request merged**